compiler/rustc_middle/src/ty/impls_ty.rs

```diff
@@ -132,35 +135,370 @@ where
     }
 }
 
-// AllocIds get resolved to whatever they point to (to be stable)
-impl<'a> HashStable<StableHashingContext<'a>> for mir::interpret::AllocId {
+impl<'a> ToStableHashKey<StableHashingContext<'a>> for region::Scope {
+    type KeyType = region::Scope;
+
+    #[inline]
+    fn to_stable_hash_key(&self, _: &StableHashingContext<'a>) -> region::Scope {
+        *self
+    }
+}
+
+impl<'a> HashStable<StableHashingContext<'a>> for mir::interpret::ConstAlloc<'_> {
     fn hash_stable(&self, hcx: &mut StableHashingContext<'a>, hasher: &mut StableHasher) {
+        self.ty.hash_stable(hcx, hasher);
         ty::tls::with_opt(|tcx| {
-            trace!("hashing {:?}", *self);
             let tcx = tcx.expect("can't hash AllocIds during hir lowering");
-            tcx.get_global_alloc(*self).hash_stable(hcx, hasher);
+            tcx.get_global_alloc(self.alloc_id).hash_stable(hcx, hasher);
         });
     }
 }
 
-// `Relocations` with default type parameters is a sorted map.
-impl<'a, Tag> HashStable<StableHashingContext<'a>> for mir::interpret::Relocations<Tag>
-where
-    Tag: HashStable<StableHashingContext<'a>>,
-{
+impl<'a> HashStable<StableHashingContext<'a>> for mir::interpret::ConstValue<'_> {
+    fn hash_stable(&self, hcx: &mut StableHashingContext<'a>, hasher: &mut StableHasher) {
+        use mir::interpret::{ConstValue, Scalar};
+        std::mem::discriminant(self).hash_stable(hcx, hasher);
+        ty::tls::with_opt(|tcx| match self {
+            ConstValue::Scalar(scalar) => {
+                std::mem::discriminant(scalar).hash_stable(hcx, hasher);
+                match scalar {
+                    Scalar::Int(int) => int.hash_stable(hcx, hasher),
+                    Scalar::Ptr(ptr, size) => {
+                        let tcx = tcx.expect("can't hash AllocIds during hir lowering");
+                        let (ptr, offset) = ptr.into_parts();
+                        tcx.get_global_alloc(ptr).hash_stable(hcx, hasher);
+                        offset.hash_stable(hcx, hasher);
+                        size.hash_stable(hcx, hasher);
```

Comment:
Is it really necessary to hash `size`? This is the pointer size of the target, so identical for all pointers. Plus hashing is expensive.

Reply:
Good point, but I'd rather implement a correct implementation of HashStable than an optimized one first. I'll leave a FIXME and get back to it.